extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/MaterializedViewUtils.java

```diff
@@ -0,0 +1,282 @@
+ * under the License.
+ */
+
+package io.druid.query.materializedview;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import io.druid.java.util.common.Intervals;
+import io.druid.java.util.common.JodaUtils;
+import io.druid.query.Query;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.FilteredAggregatorFactory;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.filter.AndDimFilter;
+import io.druid.query.filter.BoundDimFilter;
+import io.druid.query.filter.DimFilter;
+import io.druid.query.filter.InDimFilter;
+import io.druid.query.filter.IntervalDimFilter;
+import io.druid.query.filter.LikeDimFilter;
+import io.druid.query.filter.NotDimFilter;
+import io.druid.query.filter.OrDimFilter;
+import io.druid.query.filter.RegexDimFilter;
+import io.druid.query.filter.SearchQueryDimFilter;
+import io.druid.query.filter.SelectorDimFilter;
+import io.druid.query.groupby.GroupByQuery;
+import io.druid.query.timeseries.TimeseriesQuery;
+import io.druid.query.topn.TopNQuery;
+import org.joda.time.Interval;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+public class MaterializedViewUtils 
+{
+  /**
+   * extract all dimensions in query.
+   * only support TopNQuery/TimeseriesQuery/GroupByQuery
+   * 
+   * @param query
+   * @return dimensions set in query
+   */
+  public static Set<String> getRequiredFields(Query query)
+  {
+    Set<String> dimensions = new HashSet<>();
+    Set<String> dimsInFilter = getDimensionsInFilter(query.getFilter());
+    if (dimsInFilter != null) {
+      dimensions.addAll(dimsInFilter);
+    }
+    if (query instanceof TopNQuery) {
+      TopNQuery q = (TopNQuery) query;
+      dimensions.addAll(extractFieldsFromAggregations(q.getAggregatorSpecs()));
+      dimensions.add(q.getDimensionSpec().getDimension());
+    } else if (query instanceof TimeseriesQuery) {
+      TimeseriesQuery q = (TimeseriesQuery) query;
+      dimensions.addAll(extractFieldsFromAggregations(q.getAggregatorSpecs()));
+    } else if (query instanceof GroupByQuery) {
+      GroupByQuery q = (GroupByQuery) query;
+      dimensions.addAll(extractFieldsFromAggregations(q.getAggregatorSpecs()));
+      for (DimensionSpec spec : q.getDimensions()) {
+        String dim = spec.getDimension();
+        dimensions.add(dim);
+      }
+    }
+    return dimensions;
+  }
+
+  private static Set<String> extractFieldsFromAggregations(List<AggregatorFactory> aggs) 
+  {
+    Set<String> ret = new HashSet<>();
+    for (AggregatorFactory agg : aggs) {
+      if (agg instanceof FilteredAggregatorFactory) {
+        FilteredAggregatorFactory fagg = (FilteredAggregatorFactory) agg;
+        ret.addAll(getDimensionsInFilter(fagg.getFilter()));
+      }
+      ret.addAll(agg.requiredFields());
+    }
+    return ret;
+  }
+
+  private static Set<String> getDimensionsInFilter(DimFilter dimFilter)
```

Comment:
I think it's better to add a method to `DimFilter` which returns all required column names.

Reply:
I think it's up to you. If you don't want to make this PR bigger, please raise an issue for this.